### PR TITLE
server, session, variable: normalize tidb_txn_mode values

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve?

Problem Summary:

Previously the sysvar `tidb_txn_mode` permitted three values:
- optimistic
- pessimistic
- empty string (behaves the same as optimistic)

This improves the normalization of system variables to be in upper case (consistent with most other ENUM variables), and the validation function converts "" to `OPTIMISTIC`.

Because the state is cleaned up internally, the helper function `GetReadableTxnMode()` is no longer required.

Additionally, the default is changed to `PESSIMISTIC` so that it can support the use case of `SET tidb_txn_mode = DEFAULT` (users effectively think pessmistic is the default, but from a code point of view it's not). To support this, the bootstrap logic of changing the value of tidb_txn_mode to `PESSIMISTIC` if the store is tikv is inverted to be if the store `!=` tikv.

**Note:** This means there is a slight behavior change: Upgrading from 2.1 **directly** to master will change the txn-mode from optimistic to pessimistic. But if you step through any version in between it will stay as optimistic. I think this is easy enough to describe in the docs, but long term it might be better to say that direct upgrades that skip more than 1 major digit are not supported. That way we can avoid the excessive language in docs describing edge-cases (which hurt readability).

### What is changed and how it works?

What's Changed:

Code cleanup, mostly function neutral to users. Improves usability because the behavior of "" is confusing to me, and likely to other users.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- None (the code supports the upgrade case)

### Release note 

- The handling of the `tidb_txn_mode` system variable has been improved. It is recommended to set the value to `OPTIMISTIC` or `PESSIMISTIC`, and no longer the special empty string ('') value.

### CODE REVIEW

File: session/bootstrap.go [hunk 1]
```diff
@@ -1597,8 +1597,8 @@ func doDMLWorks(s Session) {
 		// Session only variable should not be inserted.
 		if v.Scope != variable.ScopeSession {
 			vVal := v.Value
-			if v.Name == variable.TiDBTxnMode && config.GetGlobalConfig().Store == "tikv" {
-				vVal = "pessimistic"
+			if v.Name == variable.TiDBTxnMode && config.GetGlobalConfig().Store != "tikv" {
+				vVal = "OPTIMISTIC"
```
Comment (xhebox) [inline session/bootstrap.go:1601]: Maybe use `DefTiDBTxnMode` directly, is it possible?

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: session/bootstrap.go (hunk 1 above)
Comment (morgo) [reply session/bootstrap.go:1601]: `DefTiDBTxnMode` is now pessimistic, but here we are saying if the store is not tikv use `OPTIMISTIC`. I can use `ast.Optimistic` if that helps?

Comment (xhebox) [reply session/bootstrap.go:1601]: Oops, I did not notice that... No, just keep it unchanged. Do we need to change the document or so?

Comment (morgo) [reply session/bootstrap.go:1601]: Yes, the user documents should change. But I am hoping we can merge the automated docs generation and incorporate it here: https://github.com/pingcap/docs/pull/5720

**Review by xhebox — APPROVED** [review]

**Review by djshow832 — COMMENTED** [review]
I'm afraid it's a big decision to change the default txn mode. @cfzjywxk

**Comment by morgo** [comment]
> I'm afraid it's a big decision to change the default txn mode. @cfzjywxk

To clarify: it's not a default change unless the user upgrades directly from 3.0 to 5.2. This is because on bootstrap clusters are set to be pessimistic mode already.

This PR is trying to reduce the confusion which is caused by having 3 states; at the cost of an unlikely (and should not be supported) upgrade path.

**Comment by morgo** [comment]
> I'm afraid it's a big decision to change the default txn mode. @cfzjywxk

To clarify: it's not a default change unless the user upgrades directly from 3.0 to 5.2. This is because on bootstrap clusters are set to be pessimistic mode already.

This PR is trying to reduce the confusion which is caused by having 3 states; at the cost of an unlikely (and should not be supported) upgrade path.

**Comment by cfzjywxk** [comment]
> > I'm afraid it's a big decision to change the default txn mode. @cfzjywxk
> 
> To clarify: it's not a default change unless the user upgrades directly from 3.0 to 5.2. This is because on bootstrap clusters are set to be pessimistic mode already.
> 
> This PR is trying to reduce the confusion which is caused by having 3 states; at the cost of an unlikely (and should not be supported) upgrade path.

@morgo 
There are some key customers who will do the cross version upgrade, we may need to discuss this behaviour change together with the bussines team and support team.

**Comment by morgo** [comment]
> @morgo
> There are some key customers who will do the cross version upgrade, we may need to discuss this behaviour change together with the bussines team and support team.

OK let me know :-)

**Pull request closed without merging**